File: kdownloader-flutter/android/src/main/kotlin/com/roxybasicneedbot/kdownloader/flutter/KDownloaderFlutterPlugin.kt
package com.roxybasicneedbot.kdownloader.flutter

import android.content.Context
import com.roxybasicneedbot.kdownloader.android.KDownloader
import com.roxybasicneedbot.kdownloader.android.persistence.entity.DownloadTaskEntity
import com.roxybasicneedbot.kdownloader.core.model.DownloadPriority
import com.roxybasicneedbot.kdownloader.core.model.DownloadRequest
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class KDownloaderFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var context: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private val scope = CoroutineScope(Dispatchers.Main)
    private var observeJob: Job? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "com.roxybasicneedbot.kdownloader/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "com.roxybasicneedbot.kdownloader/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
                observeJob = KDownloader.getInstance(context).observeAll()
                    .onEach { tasks ->
                        val serialized = tasks.map { it.toMap() }
                        events.success(serialized)
                    }
                    .launchIn(scope)
            }

            override fun onCancel(arguments: Any?) {
                observeJob?.cancel()
                observeJob = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val downloader = KDownloader.getInstance(context)
        when (call.method) {
            "enqueue" -> {
                val id = call.argument<String>("id")!!
                val url = call.argument<String>("url")!!
                val destinationDir = call.argument<String>("destinationDir")!!
                val fileName = call.argument<String>("fileName")!!
                val priorityStr = call.argument<String>("priority") ?: "NORMAL"
                val chunkCount = call.argument<Int>("chunkCount") ?: 4
                val headers = call.argument<Map<String, String>>("headers") ?: emptyMap()
                val wifiOnly = call.argument<Boolean>("wifiOnly") ?: false
                val speedLimit = call.argument<Int>("speedLimit") ?: 0
                val mirrorUrls = call.argument<List<String>>("mirrorUrls") ?: emptyList()
                val hashAlgorithm = call.argument<String>("hashAlgorithm")
                val expectedHash = call.argument<String>("expectedHash")
                val scheduleAt = call.argument<Long>("scheduleAt")
                val groupTag = call.argument<String>("groupTag")

                val request = DownloadRequest(
                    id = id,
                    url = url,
                    destinationDir = destinationDir,
                    fileName = fileName,
                    priority = DownloadPriority.valueOf(priorityStr),
                    chunkCount = chunkCount,
                    headers = headers,
                    wifiOnly = wifiOnly,
                    speedLimit = speedLimit.toLong(),
                    mirrorUrls = mirrorUrls,
                    hashAlgorithm = hashAlgorithm,
                    expectedHash = expectedHash,
                    scheduleAt = scheduleAt,
                    groupTag = groupTag
                )

                scope.launch {
                    try {
                        val taskId = downloader.enqueue(request)
                        result.success(taskId)
                    } catch (e: Exception) {
                        result.error("ENQUEUE_FAILED", e.message, null)
                    }
                }
            }
            "pause" -> {
                val id = call.argument<String>("id")!!
                scope.launch {
                    try {
                        downloader.pause(id)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("PAUSE_FAILED", e.message, null)
                    }
                }
            }
            "resume" -> {
                val id = call.argument<String>("id")!!
                scope.launch {
                    try {
                        downloader.resume(id)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("RESUME_FAILED", e.message, null)
                    }
                }
            }
            "cancel" -> {
                val id = call.argument<String>("id")!!
                scope.launch {
                    try {
                        downloader.cancel(id)
                        result.success(null)
                    } catch (e: Exception) {
                        result.error("CANCEL_FAILED", e.message, null)
                    }
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        observeJob?.cancel()
    }
}

private fun DownloadTaskEntity.toMap(): Map<String, Any?> {
    val progressMap = if (status == "DOWNLOADING" || status == "CONNECTING" || status == "PAUSED") {
        mapOf(
            "downloadedBytes" to downloadedBytes,
            "totalBytes" to totalBytes,
            "percent" to if (totalBytes > 0) ((downloadedBytes * 100) / totalBytes).toInt() else 0,
            "speedBytesPerSec" to 0,
            "speedFormatted" to "0 B/s",
            "etaSeconds" to 0,
            "etaFormatted" to "0s"
        )
    } else null

    return mapOf(
        "id" to id,
        "status" to status,
        "progress" to progressMap,
        "errorMessage" to errorMessage
    )
}

File: kdownloader-flutter/ios/Classes/KDownloaderFlutterPlugin.swift
import Flutter
import UIKit
import kdownloader_core

public class KDownloaderFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var observeTask: Task<Void, Never>?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.roxybasicneedbot.kdownloader/methods", binaryMessenger: registrar.messenger())
        let instance = KDownloaderFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        let eventChannel = FlutterEventChannel(name: "com.roxybasicneedbot.kdownloader/events", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let downloader = KDownloader.companion.instance
        
        switch call.method {
        case "enqueue":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String,
                  let url = args["url"] as? String,
                  let destinationDir = args["destinationDir"] as? String,
                  let fileName = args["fileName"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            
            // Build the priority enum if exposed to iOS
            // Assuming KDownloader_core exposes the Builder or directly DownloadRequest
            let priorityStr = args["priority"] as? String ?? "NORMAL"
            let priority: DownloadPriority = priorityStr == "HIGH" ? .high : (priorityStr == "LOW" ? .low : .normal)
            let request = DownloadRequest(
                id: id,
                url: url,
                destinationDir: destinationDir,
                fileName: fileName,
                priority: priority,
                chunkCount: Int32(args["chunkCount"] as? Int ?? 4),
                headers: args["headers"] as? [String: String] ?? [:],
                wifiOnly: args["wifiOnly"] as? Bool ?? false,
                speedLimit: args["speedLimit"] as? Int64 ?? 0,
                mirrorUrls: args["mirrorUrls"] as? [String] ?? [],
                hashAlgorithm: args["hashAlgorithm"] as? String,
                expectedHash: args["expectedHash"] as? String,
                scheduleAt: (args["scheduleAt"] as? Int64).map { KotlinLong(value: $0) },
                groupTag: args["groupTag"] as? String
            )
            
            Task {
                do {
                    let taskId = try await downloader.enqueue(request: request)
                    DispatchQueue.main.async { result(taskId) }
                } catch {
                    DispatchQueue.main.async { result(FlutterError(code: "ENQUEUE_ERROR", message: error.localizedDescription, details: nil)) }
                }
            }
            
        case "pause":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing id", details: nil))
                return
            }
            Task {
                do {
                    try await downloader.pause(id: id)
                    DispatchQueue.main.async { result(nil) }
                } catch {
                    DispatchQueue.main.async { result(FlutterError(code: "PAUSE_ERROR", message: error.localizedDescription, details: nil)) }
                }
            }
            
        case "resume":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing id", details: nil))
                return
            }
            Task {
                do {
                    try await downloader.resume(id: id)
                    DispatchQueue.main.async { result(nil) }
                } catch {
                    DispatchQueue.main.async { result(FlutterError(code: "RESUME_ERROR", message: error.localizedDescription, details: nil)) }
                }
            }
            
        case "cancel":
            guard let args = call.arguments as? [String: Any],
                  let id = args["id"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing id", details: nil))
                return
            }
            Task {
                do {
                    try await downloader.cancel(id: id)
                    DispatchQueue.main.async { result(nil) }
                } catch {
                    DispatchQueue.main.async { result(FlutterError(code: "CANCEL_ERROR", message: error.localizedDescription, details: nil)) }
                }
            }
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        
        observeTask?.cancel()
        observeTask = Task {
            do {
                // Skie bridging Flow<List<DownloadTaskEntity>> to Swift AsyncSequence
                for try await states in KDownloader.companion.instance.observeAll() {
                    let array: [[String: Any]] = states.map { task in
                        return [
                            "id": task.id,
                            "url": task.url,
                            "destinationDir": task.destinationDir,
                            "fileName": task.fileName,
                            "status": task.status,
                            "downloadedBytes": task.downloadedBytes,
                            "totalBytes": task.totalBytes,
                            "errorMessage": task.errorMessage ?? ""
                        ]
                    }
                    DispatchQueue.main.async {
                        self.eventSink?(array)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    self.eventSink?(FlutterError(code: "OBSERVE_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        observeTask?.cancel()
        self.eventSink = nil
        return nil
    }
}
